analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs

```diff
@@ -27,20 +28,47 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn
     internal class RoslynSymbolicExecution
     {
         private readonly ControlFlowGraph cfg;
+        private readonly SymbolicExecutionCheck[] checks;
 
-        public RoslynSymbolicExecution(ControlFlowGraph cfg) =>
+        public RoslynSymbolicExecution(ControlFlowGraph cfg, SymbolicExecutionCheck[] checks)
+        {
             this.cfg = cfg ?? throw new ArgumentNullException(nameof(cfg));
+            this.checks = checks ?? throw new ArgumentNullException(nameof(checks));
+            if (!checks.Any())
+            {
+                throw new ArgumentException("At least one check is expected", nameof(checks));
+            }
+        }
 
         public void Execute()
         {
-            foreach (var block in cfg.Blocks)    // ToDo: This is a temporary simplification until we support proper branching
+            foreach (var block in cfg.Blocks)   // ToDo: This is a temporary simplification until we support proper branching
             {
-                foreach(var operation in block.OperationsAndBranchValue.ToExecutionOrder()) // ToDo: This is a temporary oversimplification to scaffold the engine
+                var state = new ProgramState(); // ToDo: This is a temporary oversimplification to scaffold the enginge
+                foreach (var operation in block.OperationsAndBranchValue.ToExecutionOrder()) // ToDo: This is a temporary oversimplification to scaffold the engine
                 {
+                    state = InvokeChecks(state, (check, ps) => check.PreProcess(ps, operation));
```

Comment:
silly question: shouldn't this return a list of states?

Reply:
_In theory:_ yes.
_In practice:_ no, it's not needed by our rules.
_In the future:_ we can always upgrade if we need to.